### PR TITLE
Revise sample configs, increase memory requests, update Ray versions

### PR DESCRIPTION
**Pull request by DmitriGekhtman**
Signed-off-by: Dmitri Gekhtman <dmitri.m.gekhtman@gmail.com>

## Why are these changes needed?

This PR

- Cleans up and updates sample configs and test files
- Increments Ray versions to the latest Ray release (2.1.0) 

Memory resource requests are increased in some sample configs.
The Ray head pod has a high risk of running out of memory if it is allocated less than 2Gb memory, so that's fixed.

## Checks

- [ ] I've made sure the tests are passing. 
- Testing Strategy
   - [ ] Unit tests
   - [ ] Manual tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by kevin85421 — COMMENTED** [review]
Thank @DmitriGekhtman for this contribution! Would you mind to update:

(1) `rayVersion` and `image` for all sample YAML files in `ray-operator/config/samples`

(2) Update the following line to the latest ray release (i.e. 2.1.0)

https://github.com/ray-project/kuberay/blob/49c44bff5a807a680248b30b5563f985a075be9b/tests/framework/test_sample_raycluster_yamls.py#L57

(3) Run configuration tests on your local machine. Currently, only 3 small configuration YAMLs can be tested in KubeRay CI. We can test `ray-cluster.autoscaler.yaml` and `ray-cluster.complete.yaml` by running the following command in your local environment:

```bash
python3 test_sample_raycluster_yamls.py 2>&1 | tee log
```


https://github.com/ray-project/kuberay/blob/49c44bff5a807a680248b30b5563f985a075be9b/tests/framework/test_sample_raycluster_yamls.py#L29-L31

**Comment by kevin85421** [comment]
By the way, would you mind describing more about the relationship between increasing memory requests and OOM? Thanks!

Without the context of GKE, it is hard to understand. For me, I thought a Pod will be killed by OOM when

(1) Memory usage exceeds `memory limit`. => In this case, making `memory request` equal to `memory limit` cannot resolve OOM.

(2) When the cluster runs out of resources, Pods with memory usage (>= memory request, but <= memory limit) will be killed before those Pods with memory usage (<= memory request). => In this case, making `memory request` equal to `memory limit` will make other Pods OOM or Pods cannot be scheduled.

Only in the context of GKE autopilot, it makes sense to avoid OOM by increasing memory requests. Thanks!

**Comment by DmitriGekhtman** [comment]
Right, there are two related motivations:

1. GKE autopilot has a tendency to artificially decrease memory requests to the limits, which will cause unexpected OOMs.

2. Ray has no understanding of limits vs. requests, as Ray was designed with VMs in mind. KubeRay advertises the memory **limit** to Ray as the Ray node's memory capacity and that's all that Ray sees. Thus, for consistency of Ray's behavior, it's best to simply line up limits requests. 
Otherwise, whether your Ray pod OOMs or not is sensitive to the K8s environment in ways that are difficult to understand. 
K8s container limit == K8s container request == Ray node resource capacity is much easier to understand.

**Comment by DmitriGekhtman** [comment]
> (2) When the cluster runs out of resources, Pods with memory usage (>= memory request, but <= memory limit) will be killed before those Pods with memory usage (<= memory request). => In this case, making memory request equal to memory limit will make other Pods OOM or Pods cannot be scheduled.

In particular, if you don't set requests=limits, you're prone to unexpected results due to the behavior you've described.
You are more prone to noisy neighbors issues that depend on what your K8s nodes look like and what other workloads are running on these nodes. So when a Ray worker OOMs, you won't know if you made a mistake setting up your Ray workload or if another user was running a resource intensive application that overlapped on the same physical K8s nodes.

**Comment by DmitriGekhtman** [comment]
TODO: Run configuration tests.

I'll also add a blurb to the development docs about these tests.

**Comment by kevin85421** [comment]
The update looks good to me. Feel free to merge it if the configuration tests pass.

**Review by kevin85421 — COMMENTED** [review]
* Update the version of Ray from 2.0.0 to 2.1.0 in configuration test CI.
  https://github.com/ray-project/kuberay/blob/49c44bff5a807a680248b30b5563f985a075be9b/.github/workflows/test-job.yaml#L343-L360


* Update the default Ray image from 2.0.0 to 2.1.0 in configuration test scripts.
  * [test_sample_raycluster_yamls.py#L57](https://github.com/ray-project/kuberay/blob/49c44bff5a807a680248b30b5563f985a075be9b/tests/framework/test_sample_raycluster_yamls.py#L57)
  * [test_sample_rayservice_yamls.py#L33](https://github.com/ray-project/kuberay/blob/master/tests/framework/test_sample_rayservice_yamls.py#L33)

**Comment by DmitriGekhtman** [comment]
Need to debug some of the tests -- great to see the tests catching misconfigurations.

**Comment by DmitriGekhtman** [comment]
Still having trouble debugging the compatibility test.

File: tests/framework/prototype.py
Comment (Jeffwan) [inline tests/framework/prototype.py:509]: Do we also like to test 2.0.0 in parallel?

File: tests/framework/config/ray-cluster.mini.yaml.template
Comment (Jeffwan) [inline tests/framework/config/ray-cluster.mini.yaml.template:92]: We don't have any reference on `TYPE` now?

File: tests/framework/config/ray-cluster.mini.yaml.template
Comment (Jeffwan) [inline tests/framework/config/ray-cluster.mini.yaml.template:25]: `node-ip-address: $MY_POD_IP` . Without this config, ray core will use pod id (node ip inside container)? This should be out of box supported?

File: tests/framework/config/ray-cluster.mini.yaml.template
Comment (DmitriGekhtman) [reply tests/framework/config/ray-cluster.mini.yaml.template:92]: There are no required envs, as far as I know.

File: tests/framework/config/ray-cluster.mini.yaml.template
Comment (DmitriGekhtman) [reply tests/framework/config/ray-cluster.mini.yaml.template:25]: The ip is correctly resolved by the Ray core out of the box.

**Comment by DmitriGekhtman** [comment]
Maybe the test is broken in master?
Still hitting the Ray Client connection timeout.

**Comment by DmitriGekhtman** [comment]
Goes without saying, but the use of Ray client with a node port service is ridiculous -- these tests should use kubectl exec.

File: tests/framework/prototype.py
Comment (DmitriGekhtman) [reply tests/framework/prototype.py:509]: Eventually.
This test is a prototype.

File: tests/config/ray-cluster.mini.yaml.template
Comment (kevin85421) [inline tests/config/ray-cluster.mini.yaml.template:68]: `rayCluster: raycluster-compatibility-test` cannot be removed in this PR. It is used in the NodePort service.

https://github.com/ray-project/kuberay/blob/3a647129e248640f2314bfafd257f6b7a5db232f/tests/config/raycluster-service.yaml#L8

File: tests/config/ray-cluster.mini.yaml.template
Comment (kevin85421) [inline tests/config/ray-cluster.mini.yaml.template:33]: `rayCluster: raycluster-compatibility-test` cannot be removed in the head pod. It is used in the NodePort service.

https://github.com/ray-project/kuberay/blob/3a647129e248640f2314bfafd257f6b7a5db232f/tests/config/raycluster-service.yaml#L8

Comment (DmitriGekhtman) [reply tests/config/ray-cluster.mini.yaml.template:33]: I fixed it to use the selector used by the built-in service
https://github.com/ray-project/kuberay/pull/761/commits/f6b43d7cbd46636c61f5e0c19a46408049cfe90e

I tried to remove the extra service by using spec.headGroupSpec.ServiceType but that didn't work out.

Comment (DmitriGekhtman) [reply tests/config/ray-cluster.mini.yaml.template:33]: Actually, switched it back to how it was. Changed the label to something more descriptive and added explanation of what it's for.

**Comment by DmitriGekhtman** [comment]
Woooooo, the CI is green after https://github.com/ray-project/kuberay/pull/759
Nice work @kevin85421. Would you mind giving this PR a final review?

**Review by Jeffwan — APPROVED** [review]
most parts look good to me. I did't check test framework part. leave it to Kevein

File: tests/config/ray-service-cluster-update.yaml.template
Comment (kevin85421) [inline tests/config/ray-service-cluster-update.yaml.template:47]: This file should be deleted.

File: tests/config/ray-service-serve-update.yaml.template
Comment (kevin85421) [inline tests/config/ray-service-serve-update.yaml.template:47]: This file should be deleted.

Comment (kevin85421) [reply tests/config/ray-service-serve-update.yaml.template:47]: https://github.com/ray-project/kuberay/pull/759/files#diff-db7eeddbec6895cf05cb70312b97d42f54bdd61ac34ee18d3f74d02ab349b5bfL274

File: tests/config/ray-service-cluster-update.yaml.template
Comment (kevin85421) [reply tests/config/ray-service-cluster-update.yaml.template:47]: https://github.com/ray-project/kuberay/pull/759/files#diff-db7eeddbec6895cf05cb70312b97d42f54bdd61ac34ee18d3f74d02ab349b5bfL274

**Review by kevin85421 — APPROVED** [review]
It looks good to me, but it is hard for me to verify this big PR. The good thing is that I will check all sample YAML files before v0.4.0 release except `ray-cluster.complete.large.yaml` and `ray-cluster.autoscaler.large.yaml`. Hence, feel free to merge this PR if you double check the two large YAML files.

File: tests/config/ray-service-cluster-update.yaml.template
Comment (DmitriGekhtman) [reply tests/config/ray-service-cluster-update.yaml.template:47]: Thanks for pointing that out!

File: tests/config/ray-service-serve-update.yaml.template
Comment (DmitriGekhtman) [reply tests/config/ray-service-serve-update.yaml.template:47]: Will do!

**Comment by DmitriGekhtman** [comment]
> It looks good to me, but it is hard for me to verify this big PR. The good thing is that I will check all sample YAML files [1] before v0.4.0 release except `ray-cluster.complete.large.yaml` and `ray-cluster.autoscaler.large.yaml`. Hence, feel free to merge this PR if you double-check the two large YAML files. Maybe we can update their memory requests, cpu requests, and repliacs to enable them to be tested on your laptop.
> 

I'll double check those.
The configuration test passes!

**Comment by DmitriGekhtman** [comment]
I tested raycluster.complete.large and raycluster.autoscaler.large on GKE -- everything lgtm.

**Pull request merged**
